### ISSUE TITLE
Stateless FunctionRef: ActorRefWithContext

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#24459]
`FunctionRef`s are nice to correlate an `ActorRef` with some extra context. This would work in a simple fashion locally, however, for remote usage, a `FunctionRef` needs to be registered because the function cannot be sent remotely, so a reference to the function needs to be kept locally. This means, the creator of a `FunctionRef` needs to maintain its lifetime manually and cannot use it for adhoc purposes.

The idea is now to create a stateless ActorRef with context where the context is guaranteed to be able to be sent around using remoting. It is arguable if the context actually needs to be guaranteed to be serializable or if you could just use any object (similar as message payloads are not restricted as well).

The implementation would be entirely in serialization:

 * when sending out an ActorRefWithContext, we provide a serializer and deserializer for that
 * sending a message to an ActorRefWithContext, wraps the message in a `MessageWithContext(msg, context)` which is part of the API and for which we also define serializers and sends to the original ref
 * the original actor needs to handle the `MessageWithContext` message explicitly

**Comment by jrudolph** [comment]
Alternatives:

 * using proper actors (that though have the same lifetime issues as FunctionRefs)
 * a better managed FunctionRef that has automatic lifetime guarantees, e.g. like a max number of responses to expect, and a timeout (basically a generalization of PromiseActorRefs)

**Comment by patriknw** [comment]
Interesting idea. Crucial would be to make it very easy to include the serialized format (bytes) of the ActorRefWithContext in user messages. Ordinary ActorRef is easy since they can be represented as a String.

**Comment by jrudolph** [comment]
> Crucial would be to make it very easy to include the serialized format (bytes) of the ActorRefWithContext in user messages. Ordinary ActorRef is easy since they can be represented as a String.

Good point. I guess it's bad API design on our part that we only provide `Serialization.serializedActorPath` which already assumes that an ActorRef can (or should) always be represented as a `String`. It would have been easier if we'd provided just a byte array so we could have hidden anything behind that. We can still make it work but will find a string representation of whatever the context would be (e.g. base64) and some discriminator string to tell it apart from normal ActorRefs.

**Comment by patriknw** [comment]
yes, a base64 string isn't too bad since it makes the api compatible with ordinary ActorRef

**Comment by ktoso** [comment]
Sounds good, no problem to separate the "payload" from path with some special `$$$` etc ;)
